frame/dex-router/src/lib.rs

```diff
@@ -0,0 +1,190 @@
+
+#[frame_support::pallet]
+pub mod pallet {
+	use codec::{Codec, FullCodec};
+	use composable_traits::{
+		defi::LiftedFixedBalance,
+		dex::{CurveAmm, DexRoute, DexRouteNode, DexRouter},
+	};
+	use core::fmt::Debug;
+	use frame_support::pallet_prelude::*;
+	use sp_runtime::{
+		traits::{
+			AtLeast32BitUnsigned, CheckedAdd, CheckedMul, CheckedSub, IntegerSquareRoot, One, Zero,
+		},
+		DispatchResult, FixedPointOperand,
+	};
+
+	#[pallet::config]
+	pub trait Config: frame_system::Config {
+		type Event: From<Event<Self>> + IsType<<Self as frame_system::Config>::Event>;
+		type AssetId: FullCodec
+			+ Eq
+			+ PartialEq
+			+ Copy
+			+ MaybeSerializeDeserialize
+			+ Debug
+			+ Default
+			+ TypeInfo
+			+ Ord;
+		type Balance: Default
+			+ Parameter
+			+ Codec
+			+ Copy
+			+ Ord
+			+ CheckedAdd
+			+ CheckedSub
+			+ CheckedMul
+			+ AtLeast32BitUnsigned
+			+ From<u64> // at least 64 bit
+			+ Zero
+			+ One
+			+ IntegerSquareRoot
+			+ FixedPointOperand
+			+ Into<LiftedFixedBalance>
+			+ Into<u128>; // cannot do From<u128>, until LiftedFixedBalance integer part is larger than 128
+			  // bit
+		/// The maximum hops in the route.
+		type MaxHopsInRoute: Get<u32> + TypeInfo;
+		type PoolId: FullCodec
+			+ Default
+			+ TypeInfo
+			+ Eq
+			+ PartialEq
+			+ Ord
+			+ Copy
+			+ Debug
+			+ CheckedAdd
+			+ Zero
+			+ One;
+		type PoolTokenIndex: Copy + Debug + Eq + Into<u32>;
+		type StableSwapDex: CurveAmm<
+			AssetId = Self::AssetId,
+			Balance = Self::Balance,
+			AccountId = Self::AccountId,
+			PoolTokenIndex = Self::PoolTokenIndex,
+			PoolId = Self::PoolId,
+		>;
+		type ConstantProductDex: CurveAmm<
+			AssetId = Self::AssetId,
+			Balance = Self::Balance,
+			AccountId = Self::AccountId,
+			PoolTokenIndex = Self::PoolTokenIndex,
+			PoolId = Self::PoolId,
+		>;
+	}
+
+	#[pallet::pallet]
+	#[pallet::generate_store(trait Store)]
+	pub struct Pallet<T>(_);
+
+	#[pallet::storage]
+	pub type DexRoutes<T: Config> = StorageDoubleMap<
+		_,
+		Blake2_128Concat,
+		T::AssetId,
+		Blake2_128Concat,
+		T::AssetId,
+		DexRoute<T::PoolId, T::MaxHopsInRoute>,
+		OptionQuery,
+	>;
+
+	#[pallet::error]
+	pub enum Error<T> {
+		/// Number of hops in route exceeded maximum limit.
+		MaxHopsExceeded,
+	}
+
+	#[pallet::event]
+	#[pallet::generate_deposit(pub(super) fn deposit_event)]
+	pub enum Event<T: Config> {
+		RouteAdded {
+			x_asset_id: T::AssetId,
+			y_asset_id: T::AssetId,
+			route: Vec<DexRouteNode<T::PoolId>>,
+		},
+		RouteDeleted {
+			x_asset_id: T::AssetId,
+			y_asset_id: T::AssetId,
+			route: Vec<DexRouteNode<T::PoolId>>,
+		},
+		RouteUpdated {
+			x_asset_id: T::AssetId,
+			y_asset_id: T::AssetId,
+			route: Vec<DexRouteNode<T::PoolId>>,
+		},
+	}
+
+	#[pallet::call]
+	impl<T: Config> Pallet<T> {}
+
+	impl<T: Config> Pallet<T> {}
+
+	impl<T: Config> DexRouter<T::AssetId, T::PoolId> for Pallet<T> {
+		fn update_route(
+			asset_pair: (T::AssetId, T::AssetId),
+			route: Option<Vec<DexRouteNode<T::PoolId>>>,
```

Comment:
Consider using bounded vector. Does pallet::constat get provides constant to use as bound?

Reply:
AFAIU `BoundedVec` is only beneficial for `Substrate::Storage` , no need to use it over `Vec` in regular code.

Reply:
not in this pr, but BoundedVec documents limits and avoids more code like in codebase

``
let bounded_route =
						route_vec.clone().try_into().map_err(|_| Error::<T>::MaxHopsExceeded)?;
```
afaik we have support crate to help with that

Reply:
It is a good thing to avoid arbitrary Vec over anything we are writing as it will enforce the usage of this type up to the pallet and up to the extrinsic